library/core/src/iter/adapters/flatten.rs

```diff
@@ -444,3 +473,21 @@ where
         init
     }
 }
+
+trait ConstSizeIterable {
+    fn size() -> Option<usize>;
+}
+
+impl<T> ConstSizeIterable for T {
+    #[inline]
+    default fn size() -> Option<usize> {
+        None
+    }
+}
+
+impl<T, const N: usize> ConstSizeIterable for [T; N] {
```

Comment:
Maybe `IntoConstSizeIterator` would be a better name? `ConstSizeIterable` sounds like the implementer is already an `Iterator`, but `[T; N]` is not. (technically `Iterable` could be a synonymous of `IntoIterator`, at least in Java it has that meaning, however `InPlaceIterable` already uses it to mean an `Iterator`, so one of them should be changed)

Reply:
I was going by the plain english meaning of the *-able* suffix meaning that you can do something with it. Iterable -> it can be iterated, iterator -> it does the iterating. So InPlaceIterable means it *can* be iterated in place, not that it always does that. Also, any Iterator is iterable, so that's a superset. But yes in rust `IntoIterator` already expresses this.

But `IntoConstSizeIterator` is misleading, since it's not the Iterator that's const-sized. I think `ConstSizeIntoIterator` expresses it more clearly instead.

Reply:
`ConstSizeIntoIterator` sounds good.

> So InPlaceIterable means it can be iterated in place,

The problem I have is that by going with that logic then `ConstSizeIterable` should also mean it can be somehow be iterated/advanced, however in this case it can only be turned into an iterator. Overall I think we should either properly define what we mean with `Iterable` or just stick with `Iterator`/`IntoIterator`